modules/editor/grammar/post.pegjs

```diff
@@ -0,0 +1,95 @@
+Document
+  = WP_Block_List
+
+WP_Block_List
+  = WP_Block*
+
+WP_Block
+  = WP_Block_Balanced
+  / WP_Block_Html
+
+WP_Block_Balanced
+  = s:WP_Block_Start ts:(!WP_Block_End c:Any { return c })+ e:WP_Block_End
+  { return {
+    blockType: s.blockType,
+    attrs: s.attrs,
+    rawContent: ts.join( '' ),
+  } }
+
+WP_Block_Html
+  = ts:(!WP_Block_Balanced c:Any { return c })+
+  {
+    return {
+      blockType: 'html',
+      attrs: {},
+      rawContent: ts.join('')
+    }
+  }
+
+WP_Block_Start
+  = "<!--" __ "wp:" blockType:WP_Block_Type attrs:WP_Block_Attribute_List _? "-->"
+  { return {
+    type: 'WP_Block_Start',
+    blockType,
+    attrs,
+    text: text()
+  } }
+
+WP_Block_End
+  = "<!--" __ "/wp" __ "-->"
```

Comment:
Playing a little bit of devil's advocate: How is this any different than how the JavaScript runtime is smart enough to interpret the closing curly brace `}` of an `if` condition. Other languages might make this more explicit e.g. `fi`, `endif` but certainly doesn't seem indicative of a poorly constructed syntax.